ts/src/pro/bitpanda.ts

```diff
@@ -0,0 +1,1335 @@
+         * @param {string} timeframe the length of time each candle represents
+         * @param {int|undefined} since timestamp in ms of the earliest candle to fetch
+         * @param {int|undefined} limit the maximum amount of candles to fetch
+         * @param {object} params extra parameters specific to the bitpanda api endpoint
+         * @returns {[[int]]} A list of candles ordered as timestamp, open, high, low, close, volume
+         */
+        await this.loadMarkets ();
+        const market = this.market (symbol);
+        symbol = market['symbol'];
+        const marketId = market['id'];
+        const url = this.urls['api']['ws'];
+        const timeframes = this.safeValue (this.options, 'timeframes', {});
+        const timeframeId = this.safeValue (timeframes, timeframe);
+        if (timeframeId === undefined) {
+            throw new NotSupported (this.id + ' this interval is not supported, please provide one of the supported timeframes');
+        }
+        const messageHash = 'ohlcv.' + symbol + '.' + timeframe;
+        const subscriptionHash = 'CANDLESTICKS';
+        const client = this.safeValue (this.clients, url);
+        let type = 'SUBSCRIBE';
+        let subscription = {};
+        if (client !== undefined) {
+            subscription = this.safeValue (client.subscriptions, subscriptionHash);
+            if (subscription !== undefined) {
+                const ohlcvMarket = this.safeValue (subscription, marketId, {});
+                const marketSubscribed = this.safeValue (ohlcvMarket, timeframe, false);
+                if (!marketSubscribed) {
+                    type = 'UPDATE_SUBSCRIPTION';
+                    client.subscriptions[subscriptionHash] = undefined;
+                }
+            } else {
+                subscription = {};
+            }
+        }
+        const subscriptionMarketId = this.safeValue (subscription, marketId);
+        if (subscriptionMarketId === undefined) {
+            subscription[marketId] = {};
+        }
+        subscription[marketId][timeframe] = true;
+        const properties = [];
+        const marketIds = Object.keys (subscription);
+        for (let i = 0; i < marketIds.length; i++) {
+            const marketIdtimeframes = Object.keys (subscription[marketIds[i]]);
+            for (let ii = 0; ii < marketIdtimeframes.length; ii++) {
+                const marketTimeframeId = this.safeValue (timeframes, timeframe);
+                const property = {
+                    'instrument_code': marketIds[i],
+                    'time_granularity': marketTimeframeId,
+                };
+                properties.push (property);
+            }
+        }
+        const request = {
+            'type': type,
+            'channels': [
+                {
+                    'name': 'CANDLESTICKS',
+                    'properties': properties,
+                },
+            ],
+        };
```

Comment:
question, can't this logic be abstracted using `watchMultiple` like you did with other methods?

Reply:
I looked into it.. but as we look both at the timeframe and marketId it´s a bit different than the other watchMultiple. And seemed more complicated to have one function do everything than keep it seperate even if the code is repeated... decided to keep it KISS over DRY 🤣